### DISCUSSION TITLE
How I can use "-" symbol (minus) in url?

### DISCUSSION THREAD

**Discussion by averichev** [rwf2/Rocket#2066] [Questions]
Hello, how I can use "-" symbol (minus) in url?
`#[get("/page-<id>")]` not worked

**Comment by somehowchris** [comment]
Have you just tried it?

**Comment by the10thWiz** [comment]
`-` works just fine in urls. However, I think your issue is that url parameters must take up an entire path segment.

so `#[get("/get-<id>")]` has NO url paramters, it just responds to the literal url `/get-<id>`. What I think you want is `#[get("/get/<id>")]`, which will respond to urls such as `/get/64` (assuming 64 is a valid id).

**Reply by jhawkesworth** [reply]
/get/64 works as you say but at least with 0.5 rc2 /get/-64 results in a 404 for me. I think I will have to reorganise things bit to send negative values via the url.

**Reply by somehowchris** [reply]
@jhawkesworth really? you're sure you're using `i`-ints and not `u`-ints?

**Reply by jhawkesworth** [reply]
@somehowchris yeah I'm using an i32.  The issue in my case was my parameter guard was failing the '-' as I had made it too strict.

Here's my reproducer.  Call http://127.0.0.1:8000/api/ctof/-64 to make it 404.

```rust
/* 
Cargo.toml contains:

[dependencies]
rocket = "0.5.0-rc.2"

[dependencies.rocket_dyn_templates]
version = "0.1.0-rc.2"
features = ["handlebars"]
 */

#[macro_use] extern crate rocket;

use rocket_dyn_templates::{Template, context};
use rocket::{Build, Rocket};

use rocket::request::FromParam;

#[catch(404)] fn not_found() -> &'static str { "Nothing here, sorry!" }
#[catch(500)] fn just_500() -> &'static str { "Whoops!?" }

struct IntTemp {
    degrees: i32
}

impl<'a> FromParam<'a> for IntTemp<> {
    type Error = &'a str;
    fn from_param(param: &'a str) -> Result<Self, Self::Error> {
        param.chars().all(|c| c.is_alphanumeric())
            .then(|| IntTemp{degrees: param.parse::<i32>().unwrap()})
            .ok_or(param)
    }
}

#[get("/ctof/<celsius>")]
fn c_to_f(celsius: IntTemp) -> Option<String> {
    let converted = (celsius.degrees * 9 / 5 ) +32;
    Option::from(converted.to_string())
}

#[get("/ftoc/<farenheit>")]
fn f_to_c(farenheit: IntTemp) -> Option<String> {
    let converted = (farenheit.degrees - 32) * 5 / 9;
    Option::from(converted.to_string())
}


#[get("/")]
fn index() -> &'static str {
    "Hello, world! I am an index page"
}

#[get("/hello")]
fn hello() -> Template {
    Template::render("hello", context! { greeting: "Hello, world!  I am a rocket app", })
}

#[launch]
fn rocket() -> Rocket<Build>  {
    rocket::build()
        .register("/", catchers![not_found, just_500])
        .register("/api", catchers![not_found, just_500])
        .mount("/", routes![index])
        .mount("/api", routes![hello, c_to_f, f_to_c])
        .attach(Template::fairing())
}
```
rocket log output alerted me something was wrong with the guard as it forwards instead of returning success:
```
GET /api/ctof/64 text/html:
   >> Matched: (c_to_f) GET /api/ctof/<celsius>
   >> Outcome: Success
   >> Response succeeded.
GET /api/ctof/-64 text/html:
   >> Matched: (c_to_f) GET /api/ctof/<celsius>
   >> Parameter guard `celsius: IntTemp` is forwarding: "-64".
   >> Outcome: Forward
   >> No matching routes for GET /api/ctof/-64 text/html.
   >> Responding with registered (not_found) /api 404 catcher.
   >> Response succeeded.
```
change the FromParam impl so the '-' is accepted like so and it all works:

```rust
impl<'a> FromParam<'a> for IntTemp<> {
    type Error = &'a str;
    fn from_param(param: &'a str) -> Result<Self, Self::Error> {
        param.chars().all(|c| c.is_alphanumeric() || c.eq_ignore_ascii_case(&'-'))
            .then(|| IntTemp{degrees: param.parse::<i32>().unwrap()})
            .ok_or(param)
    }
}


```

**Reply by SergioBenitez** [reply]
@jhawkesworth It looks like the entire implementation can be `i64::parse(param).map(IntTemp)`. Why isn't it? What's more, your implementation as it stands will panic: `a` isn't a valid i32, but you pass it to i32::parse and then unwrap.

**Reply by jhawkesworth** [reply]
@SergioBenitez - thanks for this, that's a much better solution and a good point about the potential for panic.  There is plenty wrong with my code above anyway - unfriendly api, no lower bound on the conversions etc.  I couldn't quite get the .map syntax to work and have switched to f32 for the IntTemp representation, but wound up with the following which I think is a bit of an improvement.  I still have much to learn.

```rust
        let parse_result = param.parse::<f32>();
        match parse_result {
            Ok(parsed) => Ok(IntTemp { degrees: parsed }),
            Err(..) => Err(param)
        }
```
